ops/model.py

```diff
@@ -3784,32 +3716,35 @@ def relation_set(self, relation_id: int, data: Mapping[str, str], is_app: bool)
                 f'{self._juju_context.version}'
             )
 
-        args = ['relation-set', '-r', str(relation_id)]
-        if is_app:
-            args.append('--app')
-        args.extend(['--file', '-'])
-
         try:
-            content = yaml.safe_dump(data)
-            self._run(*args, input_stream=content)
-        except ModelError as e:
-            if self._is_relation_not_found(e):
+            with self._tracing_span('relation-set', relation_id, data, is_app):
+                hookcmds.relation_set(data, relation_id, app=is_app)
+        except hookcmds.Error as e:
+            self._check_for_security_event('relation-set', e.returncode, e.stderr)
+            err = ModelError(e.stderr)
+            if self._is_relation_not_found(err):
                 raise RelationNotFoundError() from e
-            raise
+            raise err from e
 
     def relation_model_get(self, relation_id: int) -> dict[str, Any]:
-        args = ['relation-model-get', '-r', str(relation_id)]
         try:
-            result = self._run(*args, return_output=True, use_json=True)
-            return typing.cast('dict[str, Any]', result)
-        except ModelError as e:
-            if self._is_relation_not_found(e):
+            with self._tracing_span('relation-model-get', relation_id):
+                raw = hookcmds.relation_model_get(relation_id)
+        except hookcmds.Error as e:
+            self._check_for_security_event('relation-model-get', e.returncode, e.stderr)
+            err = ModelError(e.stderr)
+            if self._is_relation_not_found(err):
                 raise RelationNotFoundError() from e
-            raise
+            raise err from e
+        return dataclasses.asdict(raw)
 
     def config_get(self) -> dict[str, bool | int | float | str]:
-        out = self._run('config-get', return_output=True, use_json=True)
-        return typing.cast('dict[str, bool | int | float | str]', out)
+        try:
+            with self._tracing_span('config-get'):
+                return dict(hookcmds.config_get())
```

Comment:
The `dict(...)` is awkward, when it's actually just a dict already. I wonder if we should simplify hookcmds return values and just have them return plain `dict` instead of mapping. Seems better to "take abstract types as parameters, use concrete types as return values". It's never going to be anything but a `dict` anyway.